File: vs2003_cd02/Program Files/Microsoft Visual Studio .NET 2003/SDK/v1.1/QuickStart/winforms/samples/data/update/cs/customersdataset/CustomersDataSet.cs
//-----------------------------------------------------------------------
//  This file is part of the Microsoft .NET SDK Code Samples.
// 
//  Copyright (C) Microsoft Corporation.  All rights reserved.
// 
//This source code is intended only as a supplement to Microsoft
//Development Tools and/or on-line documentation.  See these other
//materials for detailed information regarding Microsoft code samples.
// 
//THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
//KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
//IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
//PARTICULAR PURPOSE.
//-----------------------------------------------------------------------

namespace Microsoft.Samples.Windows.Forms.Cs.Update.Data {
    using System;
    using System.Data;
    using Microsoft.Win32;
    
    
    public class CustomersDataSet : System.Data.DataSet {
        
        private CustomersDataTable tableCustomers;
        
        public CustomersDataSet() {
            this.InitClass();
        }
        
        [System.ComponentModel.DesignerSerializationVisibilityAttribute(System.ComponentModel.DesignerSerializationVisibility.Content)]
        public CustomersDataTable Customers {
            get {
                return this.tableCustomers;
            }
        }
        
        private void InitClass() {
            this.DataSetName = "CustomersDataSet";
            this.Namespace = "";
            this.tableCustomers = new CustomersDataTable();
            this.Tables.Add(this.tableCustomers);
        }
        
        public delegate void CustomersRowChangeEventHandler(object sender, CustomersRowChangeEvent e);
        
        public class CustomersDataTable : DataTable, System.Collections.IEnumerable {
            
            private DataColumn columnCustomerID;
            
            private DataColumn columnCompanyName;
            
            private DataColumn columnContactName;
            
            private DataColumn columnContactTitle;
            
            private DataColumn columnAddress;
            
            private DataColumn columnCity;
            
            private DataColumn columnRegion;
            
            private DataColumn columnPostalCode;
            
            private DataColumn columnCountry;
            
            private DataColumn columnPhone;
            
            private DataColumn columnFax;
            
            internal CustomersDataTable() : 
                    base("Customers") {
                this.InitClass();
            }
            
            public int Count {
                get {
                    return this.Rows.Count;
                }
            }
            
            internal DataColumn CustomerIDColumn {
                get {
                    return this.columnCustomerID;
                }
            }
            
            internal DataColumn CompanyNameColumn {
                get {
                    return this.columnCompanyName;
                }
            }
            
            internal DataColumn ContactNameColumn {
                get {
                    return this.columnContactName;
                }
            }
            
            internal DataColumn ContactTitleColumn {
                get {
                    return this.columnContactTitle;
                }
            }
            
            internal DataColumn AddressColumn {
                get {
                    return this.columnAddress;
                }
            }
            
            internal DataColumn CityColumn {
                get {
                    return this.columnCity;
                }
            }
            
            internal DataColumn RegionColumn {
                get {
                    return this.columnRegion;
                }
            }
            
            internal DataColumn PostalCodeColumn {
                get {
                    return this.columnPostalCode;
                }
            }
            
            internal DataColumn CountryColumn {
                get {
                    return this.columnCountry;
                }
            }
            
            internal DataColumn PhoneColumn {
                get {
                    return this.columnPhone;
                }
            }
            
            internal DataColumn FaxColumn {
                get {
                    return this.columnFax;
                }
            }
            
            public CustomersRow this[int index] {
                get {
                    return ((CustomersRow)(this.Rows[index]));
                }
            }
            
            public event CustomersRowChangeEventHandler CustomersRowChanged;
            
            public event CustomersRowChangeEventHandler CustomersRowChanging;
            
            public event CustomersRowChangeEventHandler CustomersRowDeleted;
            
            public event CustomersRowChangeEventHandler CustomersRowDeleting;
            
            public void AddCustomersRow(CustomersRow row) {
                this.Rows.Add(row);
            }
            
            public CustomersRow AddCustomersRow(string CustomerID, string CompanyName, string ContactName, string ContactTitle, string Address, string City, string Region, string PostalCode, string Country, string Phone, string Fax) {
                CustomersRow rowCustomersRow = ((CustomersRow)(this.NewRow()));
                rowCustomersRow.ItemArray = new Object[] {CustomerID,
                        CompanyName,
                        ContactName,
                        ContactTitle,
                        Address,
                        City,
                        Region,
                        PostalCode,
                        Country,
                        Phone,
                        Fax};
                this.Rows.Add(rowCustomersRow);
                return rowCustomersRow;
            }
            
            public CustomersRow FindByCustomerID(string CustomerID) {
                return ((CustomersRow)(this.Rows.Find(new Object[] {CustomerID})));
            }
            
            public System.Collections.IEnumerator GetEnumerator() {
                return this.Rows.GetEnumerator();
            }
            
            private void InitClass() {
                this.columnCustomerID = new DataColumn("CustomerID", typeof(string), "", MappingType.Element);
                this.columnCustomerID.AllowDBNull = false;
                this.columnCustomerID.Unique = true;
                this.Columns.Add(this.columnCustomerID);
                this.columnCompanyName = new DataColumn("CompanyName", typeof(string), "", MappingType.Element);
                this.columnCompanyName.AllowDBNull = false;
                this.Columns.Add(this.columnCompanyName);
                this.columnContactName = new DataColumn("ContactName", typeof(string), "", MappingType.Element);
                this.Columns.Add(this.columnContactName);
                this.columnContactTitle = new DataColumn("ContactTitle", typeof(string), "", MappingType.Element);
                this.Columns.Add(this.columnContactTitle);
                this.columnAddress = new DataColumn("Address", typeof(string), "", MappingType.Element);
                this.Columns.Add(this.columnAddress);
                this.columnCity = new DataColumn("City", typeof(string), "", MappingType.Element);
                this.Columns.Add(this.columnCity);
                this.columnRegion = new DataColumn("Region", typeof(string), "", MappingType.Element);
                this.Columns.Add(this.columnRegion);
                this.columnPostalCode = new DataColumn("PostalCode", typeof(string), "", MappingType.Element);
                this.Columns.Add(this.columnPostalCode);
                this.columnCountry = new DataColumn("Country", typeof(string), "", MappingType.Element);
                this.Columns.Add(this.columnCountry);
                this.columnPhone = new DataColumn("Phone", typeof(string), "", MappingType.Element);
                this.Columns.Add(this.columnPhone);
                this.columnFax = new DataColumn("Fax", typeof(string), "", MappingType.Element);
                this.Columns.Add(this.columnFax);
                this.PrimaryKey = new DataColumn[] {this.columnCustomerID};
            }
            
            public CustomersRow NewCustomersRow() {
                return ((CustomersRow)(this.NewRow()));
            }
            
            protected override DataRow NewRowFromBuilder(DataRowBuilder builder) {
                // We need to ensure that all Rows in the tabled are typed rows.
                // Table calls newRow whenever it needs to create a row.
                // So the following conditions are covered by Row newRow(Record record)
                // * Cursor calls table.addRecord(record) 
                // * table.addRow(object[] values) calls newRow(record)    
                return new CustomersRow(builder);
            }
            
            protected override System.Type GetRowType() {
                return typeof(CustomersRow);
            }
            
            protected override void OnRowChanged(DataRowChangeEventArgs e) {
                base.OnRowChanged(e);
                if (((this.CustomersRowChanged) != (null))) {
                    this.CustomersRowChanged(this, new CustomersRowChangeEvent(((CustomersRow)(e.Row)), e.Action));
                }
            }
            
            protected override void OnRowChanging(DataRowChangeEventArgs e) {
                base.OnRowChanging(e);
                if (((this.CustomersRowChanging) != (null))) {
                    this.CustomersRowChanging(this, new CustomersRowChangeEvent(((CustomersRow)(e.Row)), e.Action));
                }
            }
            
            protected override void OnRowDeleted(DataRowChangeEventArgs e) {
                base.OnRowDeleted(e);
                if (((this.CustomersRowDeleted) != (null))) {
                    this.CustomersRowDeleted(this, new CustomersRowChangeEvent(((CustomersRow)(e.Row)), e.Action));
                }
            }
            
            protected override void OnRowDeleting(DataRowChangeEventArgs e) {
                base.OnRowDeleting(e);
                if (((this.CustomersRowDeleting) != (null))) {
                    this.CustomersRowDeleting(this, new CustomersRowChangeEvent(((CustomersRow)(e.Row)), e.Action));
                }
            }
            
            public void RemoveCustomersRow(CustomersRow row) {
                this.Rows.Remove(row);
            }
        }
        
        public class CustomersRow : DataRow {
            
            private CustomersDataTable tableCustomers;
            
            internal CustomersRow(DataRowBuilder rb) : 
                    base(rb) {
                this.tableCustomers = ((CustomersDataTable)(this.Table));
            }
            
            public string CustomerID {
                get {
                    return ((string)(this[this.tableCustomers.CustomerIDColumn]));
                }
                set {
                    this[this.tableCustomers.CustomerIDColumn] = value;
                }
            }
            
            public string CompanyName {
                get {
                    return ((string)(this[this.tableCustomers.CompanyNameColumn]));
                }
                set {
                    this[this.tableCustomers.CompanyNameColumn] = value;
                }
            }
            
            public string ContactName {
                get {
                    try {
                        return ((string)(this[this.tableCustomers.ContactNameColumn]));
                    }
                    catch (InvalidCastException e) {
                        throw new StrongTypingException("Cannot get value because it is DBNull.",e);
                    }
                }
                set {
                    this[this.tableCustomers.ContactNameColumn] = value;
                }
            }
            
            public string ContactTitle {
                get {
                    try {
                        return ((string)(this[this.tableCustomers.ContactTitleColumn]));
                    }
                    catch (InvalidCastException e) {
                        throw new StrongTypingException("Cannot get value because it is DBNull.",e);
                    }
                }
                set {
                    this[this.tableCustomers.ContactTitleColumn] = value;
                }
            }
            
            public string Address {
                get {
                    try {
                        return ((string)(this[this.tableCustomers.AddressColumn]));
                    }
                    catch (InvalidCastException e) {
                        throw new StrongTypingException("Cannot get value because it is DBNull.",e);
                    }
                }
                set {
                    this[this.tableCustomers.AddressColumn] = value;
                }
            }
            
            public string City {
                get {
                    try {
                        return ((string)(this[this.tableCustomers.CityColumn]));
                    }
                    catch (InvalidCastException e) {
                        throw new StrongTypingException("Cannot get value because it is DBNull.",e);
                    }
                }
                set {
                    this[this.tableCustomers.CityColumn] = value;
                }
            }
            
            public string Region {
                get {
                    try {
                        return ((string)(this[this.tableCustomers.RegionColumn]));
                    }
                    catch (InvalidCastException e) {
                        throw new StrongTypingException("Cannot get value because it is DBNull.",e);
                    }
                }
                set {
                    this[this.tableCustomers.RegionColumn] = value;
                }
            }
            
            public string PostalCode {
                get {
                    try {
                        return ((string)(this[this.tableCustomers.PostalCodeColumn]));
                    }
                    catch (InvalidCastException e) {
                        throw new StrongTypingException("Cannot get value because it is DBNull.",e);
                    }
                }
                set {
                    this[this.tableCustomers.PostalCodeColumn] = value;
                }
            }
            
            public string Country {
                get {
                    try {
                        return ((string)(this[this.tableCustomers.CountryColumn]));
                    }
                    catch (InvalidCastException e) {
                        throw new StrongTypingException("Cannot get value because it is DBNull.",e);
                    }
                }
                set {
                    this[this.tableCustomers.CountryColumn] = value;
                }
            }
            
            public string Phone {
                get {
                    try {
                        return ((string)(this[this.tableCustomers.PhoneColumn]));
                    }
                    catch (InvalidCastException e) {
                        throw new StrongTypingException("Cannot get value because it is DBNull.",e);
                    }
                }
                set {
                    this[this.tableCustomers.PhoneColumn] = value;
                }
            }
            
            public string Fax {
                get {
                    try {
                        return ((string)(this[this.tableCustomers.FaxColumn]));
                    }
                    catch (InvalidCastException e) {
                        throw new StrongTypingException("Cannot get value because it is DBNull.",e);
                    }
                }
                set {
                    this[this.tableCustomers.FaxColumn] = value;
                }
            }
            
            public bool IsContactNameNull() {
                return this.IsNull(this.tableCustomers.ContactNameColumn);
            }
            
            public void SetContactNameNull() {
                this[this.tableCustomers.ContactNameColumn] = Convert.DBNull;
            }
            
            public bool IsContactTitleNull() {
                return this.IsNull(this.tableCustomers.ContactTitleColumn);
            }
            
            public void SetContactTitleNull() {
                this[this.tableCustomers.ContactTitleColumn] = Convert.DBNull;
            }
            
            public bool IsAddressNull() {
                return this.IsNull(this.tableCustomers.AddressColumn);
            }
            
            public void SetAddressNull() {
                this[this.tableCustomers.AddressColumn] = Convert.DBNull;
            }
            
            public bool IsCityNull() {
                return this.IsNull(this.tableCustomers.CityColumn);
            }
            
            public void SetCityNull() {
                this[this.tableCustomers.CityColumn] = Convert.DBNull;
            }
            
            public bool IsRegionNull() {
                return this.IsNull(this.tableCustomers.RegionColumn);
            }
            
            public void SetRegionNull() {
                this[this.tableCustomers.RegionColumn] = Convert.DBNull;
            }
            
            public bool IsPostalCodeNull() {
                return this.IsNull(this.tableCustomers.PostalCodeColumn);
            }
            
            public void SetPostalCodeNull() {
                this[this.tableCustomers.PostalCodeColumn] = Convert.DBNull;
            }
            
            public bool IsCountryNull() {
                return this.IsNull(this.tableCustomers.CountryColumn);
            }
            
            public void SetCountryNull() {
                this[this.tableCustomers.CountryColumn] = Convert.DBNull;
            }
            
            public bool IsPhoneNull() {
                return this.IsNull(this.tableCustomers.PhoneColumn);
            }
            
            public void SetPhoneNull() {
                this[this.tableCustomers.PhoneColumn] = Convert.DBNull;
            }
            
            public bool IsFaxNull() {
                return this.IsNull(this.tableCustomers.FaxColumn);
            }
            
            public void SetFaxNull() {
                this[this.tableCustomers.FaxColumn] = Convert.DBNull;
            }
        }
        
        public class CustomersRowChangeEvent : EventArgs {
            
            private CustomersRow eventRow;
            
            private System.Data.DataRowAction eventAction;
            
            public CustomersRowChangeEvent(CustomersRow row, DataRowAction action) {
                this.eventRow = row;
                this.eventAction = action;
            }
            
            public CustomersRow Row {
                get {
                    return this.eventRow;
                }
            }
            
            public DataRowAction Action {
                get {
                    return this.eventAction;
                }
            }
        }
    }
}

File: vs2003_cd02/Program Files/Microsoft Visual Studio .NET 2003/SDK/v1.1/QuickStart/winforms/samples/data/update/vb/customersdataset/CustomersDataSet.vb
'-----------------------------------------------------------------------
'  This file is part of the Microsoft .NET SDK Code Samples.
' 
'  Copyright (C) Microsoft Corporation.  All rights reserved.
' 
'This source code is intended only as a supplement to Microsoft
'Development Tools and/or on-line documentation.  See these other
'materials for detailed information regarding Microsoft code samples.
' 
'THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
'KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
'IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
'PARTICULAR PURPOSE.
'-----------------------------------------------------------------------

Imports System
Imports System.Data
Imports Microsoft.Win32

Namespace Microsoft.Samples.Windows.Forms.Vb.Update.Data

    Public Class CustomersDataSet
        Inherits System.Data.DataSet

        Private tableCustomers As CustomersDataTable

        Public Sub New()
            MyBase.New()
            Me.InitClass()
        End Sub

        <System.ComponentModel.DesignerSerializationVisibilityAttribute(System.ComponentModel.DesignerSerializationVisibility.Content)> _
        Public ReadOnly Property Customers() As CustomersDataTable
            Get
                Return Me.tableCustomers
            End Get
        End Property

        Private Sub InitClass()
            Me.DataSetName = "CustomersDataSet"
            Me.Namespace = ""
            Me.tableCustomers = New CustomersDataTable()
            Me.Tables.Add(Me.tableCustomers)
        End Sub

        Public Delegate Sub CustomersRowChangeEventHandler(ByVal sender As Object, ByVal e As CustomersRowChangeEvent)

        Public Class CustomersDataTable
            Inherits DataTable
            Implements System.Collections.IEnumerable

            Private columnCustomerID As DataColumn

            Private columnCompanyName As DataColumn

            Private columnContactName As DataColumn

            Private columnContactTitle As DataColumn

            Private columnAddress As DataColumn

            Private columnCity As DataColumn

            Private column_Region As DataColumn

            Private columnPostalCode As DataColumn

            Private columnCountry As DataColumn

            Private columnPhone As DataColumn

            Private columnFax As DataColumn

            Friend Sub New()
                MyBase.New("Customers")
                Me.InitClass()
            End Sub

            Public ReadOnly Property Count() As Integer
                Get
                    Return Me.Rows.Count
                End Get
            End Property

            Friend ReadOnly Property CustomerIDColumn() As DataColumn
                Get
                    Return Me.columnCustomerID
                End Get
            End Property

            Friend ReadOnly Property CompanyNameColumn() As DataColumn
                Get
                    Return Me.columnCompanyName
                End Get
            End Property

            Friend ReadOnly Property ContactNameColumn() As DataColumn
                Get
                    Return Me.columnContactName
                End Get
            End Property

            Friend ReadOnly Property ContactTitleColumn() As DataColumn
                Get
                    Return Me.columnContactTitle
                End Get
            End Property

            Friend ReadOnly Property AddressColumn() As DataColumn
                Get
                    Return Me.columnAddress
                End Get
            End Property

            Friend ReadOnly Property CityColumn() As DataColumn
                Get
                    Return Me.columnCity
                End Get
            End Property

            Friend ReadOnly Property _RegionColumn() As DataColumn
                Get
                    Return Me.column_Region
                End Get
            End Property

            Friend ReadOnly Property PostalCodeColumn() As DataColumn
                Get
                    Return Me.columnPostalCode
                End Get
            End Property

            Friend ReadOnly Property CountryColumn() As DataColumn
                Get
                    Return Me.columnCountry
                End Get
            End Property

            Friend ReadOnly Property PhoneColumn() As DataColumn
                Get
                    Return Me.columnPhone
                End Get
            End Property

            Friend ReadOnly Property FaxColumn() As DataColumn
                Get
                    Return Me.columnFax
                End Get
            End Property

            Default Public ReadOnly Property Item(ByVal index As Integer) As CustomersRow
                Get
                    Return CType(Me.Rows(index), CustomersRow)
                End Get
            End Property

            Public Event CustomersRowChanged As CustomersRowChangeEventHandler

            Public Event CustomersRowChanging As CustomersRowChangeEventHandler

            Public Event CustomersRowDeleted As CustomersRowChangeEventHandler

            Public Event CustomersRowDeleting As CustomersRowChangeEventHandler

            Public Overloads Sub AddCustomersRow(ByVal row As CustomersRow)
                Me.Rows.Add(row)
            End Sub

            Public Overloads Function AddCustomersRow(ByVal CustomerID As String, ByVal CompanyName As String, ByVal ContactName As String, ByVal ContactTitle As String, ByVal Address As String, ByVal City As String, ByVal _Region As String, ByVal PostalCode As String, ByVal Country As String, ByVal Phone As String, ByVal Fax As String) As CustomersRow
                Dim rowCustomersRow As CustomersRow = CType(Me.NewRow, CustomersRow)
                rowCustomersRow.ItemArray = New Object() {CustomerID, CompanyName, ContactName, ContactTitle, Address, City, _Region, PostalCode, Country, Phone, Fax}
                Me.Rows.Add(rowCustomersRow)
                Return rowCustomersRow
            End Function

            Public Function FindByCustomerID(ByVal CustomerID As String) As CustomersRow
                Return CType(Me.Rows.Find(New Object() {CustomerID}), CustomersRow)
            End Function

            Public Function GetEnumerator() As System.Collections.IEnumerator Implements System.Collections.IEnumerable.GetEnumerator
                Return Me.Rows.GetEnumerator
            End Function

            Private Sub InitClass()
                Me.columnCustomerID = New DataColumn("CustomerID", GetType(System.String), "", MappingType.Element)
                Me.columnCustomerID.AllowDBNull = False
                Me.columnCustomerID.Unique = True
                Me.Columns.Add(Me.columnCustomerID)
                Me.columnCompanyName = New DataColumn("CompanyName", GetType(System.String), "", MappingType.Element)
                Me.columnCompanyName.AllowDBNull = False
                Me.Columns.Add(Me.columnCompanyName)
                Me.columnContactName = New DataColumn("ContactName", GetType(System.String), "", MappingType.Element)
                Me.Columns.Add(Me.columnContactName)
                Me.columnContactTitle = New DataColumn("ContactTitle", GetType(System.String), "", MappingType.Element)
                Me.Columns.Add(Me.columnContactTitle)
                Me.columnAddress = New DataColumn("Address", GetType(System.String), "", MappingType.Element)
                Me.Columns.Add(Me.columnAddress)
                Me.columnCity = New DataColumn("City", GetType(System.String), "", MappingType.Element)
                Me.Columns.Add(Me.columnCity)
                Me.column_Region = New DataColumn("Region", GetType(System.String), "", MappingType.Element)
                Me.Columns.Add(Me.column_Region)
                Me.columnPostalCode = New DataColumn("PostalCode", GetType(System.String), "", MappingType.Element)
                Me.Columns.Add(Me.columnPostalCode)
                Me.columnCountry = New DataColumn("Country", GetType(System.String), "", MappingType.Element)
                Me.Columns.Add(Me.columnCountry)
                Me.columnPhone = New DataColumn("Phone", GetType(System.String), "", MappingType.Element)
                Me.Columns.Add(Me.columnPhone)
                Me.columnFax = New DataColumn("Fax", GetType(System.String), "", MappingType.Element)
                Me.Columns.Add(Me.columnFax)
                Me.PrimaryKey = New DataColumn() {Me.columnCustomerID}
            End Sub

            Public Function NewCustomersRow() As CustomersRow
                Return CType(Me.NewRow, CustomersRow)
            End Function

            Protected Overrides Function NewRowFromBuilder(ByVal builder As DataRowBuilder) As DataRow
                'We need to ensure that all Rows in the tabled are typed rows.
                'Table calls newRow whenever it needs to create a row.
                'So the following conditions are covered by Row newRow(Record record)
                '* Cursor calls table.addRecord(record) 
                '* table.addRow(object[] values) calls newRow(record)    
                Return New CustomersRow(builder)
            End Function

            Protected Overrides Function GetRowType() As System.Type
                Return GetType(CustomersRow)
            End Function

            Protected Overrides Sub OnRowChanged(ByVal e As DataRowChangeEventArgs)
                MyBase.OnRowChanged(e)
                If (Not (Me.CustomersRowChangedEvent) Is Nothing) Then
                    RaiseEvent CustomersRowChanged(Me, New CustomersRowChangeEvent(CType(e.Row, CustomersRow), e.Action))
                End If
            End Sub

            Protected Overrides Sub OnRowChanging(ByVal e As DataRowChangeEventArgs)
                MyBase.OnRowChanging(e)
                If (Not (Me.CustomersRowChangingEvent) Is Nothing) Then
                    RaiseEvent CustomersRowChanging(Me, New CustomersRowChangeEvent(CType(e.Row, CustomersRow), e.Action))
                End If
            End Sub

            Protected Overrides Sub OnRowDeleted(ByVal e As DataRowChangeEventArgs)
                MyBase.OnRowDeleted(e)
                If (Not (Me.CustomersRowDeletedEvent) Is Nothing) Then
                    RaiseEvent CustomersRowDeleted(Me, New CustomersRowChangeEvent(CType(e.Row, CustomersRow), e.Action))
                End If
            End Sub

            Protected Overrides Sub OnRowDeleting(ByVal e As DataRowChangeEventArgs)
                MyBase.OnRowDeleting(e)
                If (Not (Me.CustomersRowDeletingEvent) Is Nothing) Then
                    RaiseEvent CustomersRowDeleting(Me, New CustomersRowChangeEvent(CType(e.Row, CustomersRow), e.Action))
                End If
            End Sub

            Public Sub RemoveCustomersRow(ByVal row As CustomersRow)
                Me.Rows.Remove(row)
            End Sub
        End Class

        Public Class CustomersRow
            Inherits DataRow

            Private tableCustomers As CustomersDataTable

            Friend Sub New(ByVal rb As DataRowBuilder)
                MyBase.New(rb)
                Me.tableCustomers = CType(Me.Table, CustomersDataTable)
            End Sub

            Public Property CustomerID() As String
                Get
                    Return CType(Me(Me.tableCustomers.CustomerIDColumn), String)
                End Get
                Set(ByVal Value As String)
                    Me(Me.tableCustomers.CustomerIDColumn) = Value
                End Set
            End Property

            Public Property CompanyName() As String
                Get
                    Return CType(Me(Me.tableCustomers.CompanyNameColumn), String)
                End Get
                Set(ByVal Value As String)
                    Me(Me.tableCustomers.CompanyNameColumn) = Value
                End Set
            End Property

            Public Property ContactName() As String
                Get
                    Try
                        Return CType(Me(Me.tableCustomers.ContactNameColumn), String)
                    Catch e As InvalidCastException
                        Throw New StrongTypingException("Cannot get value because it is DBNull.", e)
                    End Try
                End Get
                Set(ByVal Value As String)
                    Me(Me.tableCustomers.ContactNameColumn) = Value
                End Set
            End Property

            Public Property ContactTitle() As String
                Get
                    Try
                        Return CType(Me(Me.tableCustomers.ContactTitleColumn), String)
                    Catch e As InvalidCastException
                        Throw New StrongTypingException("Cannot get value because it is DBNull.", e)
                    End Try
                End Get
                Set(ByVal Value As String)
                    Me(Me.tableCustomers.ContactTitleColumn) = Value
                End Set
            End Property

            Public Property Address() As String
                Get
                    Try
                        Return CType(Me(Me.tableCustomers.AddressColumn), String)
                    Catch e As InvalidCastException
                        Throw New StrongTypingException("Cannot get value because it is DBNull.", e)
                    End Try
                End Get
                Set(ByVal Value As String)
                    Me(Me.tableCustomers.AddressColumn) = Value
                End Set
            End Property

            Public Property City() As String
                Get
                    Try
                        Return CType(Me(Me.tableCustomers.CityColumn), String)
                    Catch e As InvalidCastException
                        Throw New StrongTypingException("Cannot get value because it is DBNull.", e)
                    End Try
                End Get
                Set(ByVal Value As String)
                    Me(Me.tableCustomers.CityColumn) = Value
                End Set
            End Property

            Public Property _Region() As String
                Get
                    Try
                        Return CType(Me(Me.tableCustomers._RegionColumn), String)
                    Catch e As InvalidCastException
                        Throw New StrongTypingException("Cannot get value because it is DBNull.", e)
                    End Try
                End Get
                Set(ByVal Value As String)
                    Me(Me.tableCustomers._RegionColumn) = Value
                End Set
            End Property

            Public Property PostalCode() As String
                Get
                    Try
                        Return CType(Me(Me.tableCustomers.PostalCodeColumn), String)
                    Catch e As InvalidCastException
                        Throw New StrongTypingException("Cannot get value because it is DBNull.", e)
                    End Try
                End Get
                Set(ByVal Value As String)
                    Me(Me.tableCustomers.PostalCodeColumn) = Value
                End Set
            End Property

            Public Property Country() As String
                Get
                    Try
                        Return CType(Me(Me.tableCustomers.CountryColumn), String)
                    Catch e As InvalidCastException
                        Throw New StrongTypingException("Cannot get value because it is DBNull.", e)
                    End Try
                End Get
                Set(ByVal Value As String)
                    Me(Me.tableCustomers.CountryColumn) = Value
                End Set
            End Property

            Public Property Phone() As String
                Get
                    Try
                        Return CType(Me(Me.tableCustomers.PhoneColumn), String)
                    Catch e As InvalidCastException
                        Throw New StrongTypingException("Cannot get value because it is DBNull.", e)
                    End Try
                End Get
                Set(ByVal Value As String)
                    Me(Me.tableCustomers.PhoneColumn) = Value
                End Set
            End Property

            Public Property Fax() As String
                Get
                    Try
                        Return CType(Me(Me.tableCustomers.FaxColumn), String)
                    Catch e As InvalidCastException
                        Throw New StrongTypingException("Cannot get value because it is DBNull.", e)
                    End Try
                End Get
                Set(ByVal Value As String)
                    Me(Me.tableCustomers.FaxColumn) = Value
                End Set
            End Property

            Public Function IsContactNameNull() As Boolean
                Return Me.IsNull(Me.tableCustomers.ContactNameColumn)
            End Function

            Public Sub SetContactNameNull()
                Me(Me.tableCustomers.ContactNameColumn) = Convert.DBNull
            End Sub

            Public Function IsContactTitleNull() As Boolean
                Return Me.IsNull(Me.tableCustomers.ContactTitleColumn)
            End Function

            Public Sub SetContactTitleNull()
                Me(Me.tableCustomers.ContactTitleColumn) = Convert.DBNull
            End Sub

            Public Function IsAddressNull() As Boolean
                Return Me.IsNull(Me.tableCustomers.AddressColumn)
            End Function

            Public Sub SetAddressNull()
                Me(Me.tableCustomers.AddressColumn) = Convert.DBNull
            End Sub

            Public Function IsCityNull() As Boolean
                Return Me.IsNull(Me.tableCustomers.CityColumn)
            End Function

            Public Sub SetCityNull()
                Me(Me.tableCustomers.CityColumn) = Convert.DBNull
            End Sub

            Public Function Is_RegionNull() As Boolean
                Return Me.IsNull(Me.tableCustomers._RegionColumn)
            End Function

            Public Sub Set_RegionNull()
                Me(Me.tableCustomers._RegionColumn) = Convert.DBNull
            End Sub

            Public Function IsPostalCodeNull() As Boolean
                Return Me.IsNull(Me.tableCustomers.PostalCodeColumn)
            End Function

            Public Sub SetPostalCodeNull()
                Me(Me.tableCustomers.PostalCodeColumn) = Convert.DBNull
            End Sub

            Public Function IsCountryNull() As Boolean
                Return Me.IsNull(Me.tableCustomers.CountryColumn)
            End Function

            Public Sub SetCountryNull()
                Me(Me.tableCustomers.CountryColumn) = Convert.DBNull
            End Sub

            Public Function IsPhoneNull() As Boolean
                Return Me.IsNull(Me.tableCustomers.PhoneColumn)
            End Function

            Public Sub SetPhoneNull()
                Me(Me.tableCustomers.PhoneColumn) = Convert.DBNull
            End Sub

            Public Function IsFaxNull() As Boolean
                Return Me.IsNull(Me.tableCustomers.FaxColumn)
            End Function

            Public Sub SetFaxNull()
                Me(Me.tableCustomers.FaxColumn) = Convert.DBNull
            End Sub
        End Class

        Public Class CustomersRowChangeEvent
            Inherits EventArgs

            Private eventRow As CustomersRow

            Private eventAction As System.Data.DataRowAction

            Public Sub New(ByVal row As CustomersRow, ByVal action As DataRowAction)
                MyBase.New()
                Me.eventRow = row
                Me.eventAction = action
            End Sub

            Public ReadOnly Property Row() As CustomersRow
                Get
                    Return Me.eventRow
                End Get
            End Property

            Public ReadOnly Property Action() As DataRowAction
                Get
                    Return Me.eventAction
                End Get
            End Property
        End Class
    End Class
End Namespace